src/saving/ISaveUpgradeStep.cs

```diff
@@ -89,6 +90,82 @@ protected override void CheckAndUpdateProperty(JProperty property)
         }
     }
 
+    internal class UpgradeStep055To056 : BaseRecursiveJSONWalkerStep
```

Comment:
I might need some help with that.
This code does not work, it crashes with an `AggregateException`: `Collection was modified; enumeration operation may not execute` on line `ISaveUpgradeStep.cs:181`.

Reply:
Look at how I did it for the json being contained in the dictionary keys. Basically you need to make a list of the things you want to update and loop that list, that way you aren't modifying the structure you are looping through. Have a look at `DetectAndUpdateKeysThatAreJSON`

Reply:
I was going to say that it doesn't make super ton of sense to move the behaviour values out of the species object to a separate composite object, that just seems unnecessary to me. But if you guys want that change, I can accept it.
I'd personally just make a class for holding the behaviour values while in the editor and not update the species to use that.

Reply:
I can see the point that a dictionary is easier to change in the future. Also, changing the storage in the species goes along well with the editor storage.

Reply:
Actually, are they stored with string keys?
If they are that yells a huge design problem to me.
I thought you had made based on adQuid's comment something like:
```
public class BehaviourValues {
    public float Aggression ...etc
}
```

I don't want random properties being stored as behaviour values in the species.

Reply:
The dictionary is based on enum keys, not string keys, which is what I was suggesting: https://github.com/Revolutionary-Games/Thrive/pull/2617/commits/5a78125163e86bf8cc0526b1fea4da0ee6cbd83a#diff-3246008dccf5d29ed6d40ffae67cd0c23d2e98886134ce85d4096611b899be6aR31

When serialized, of course, these are presumably stored as strings, which someone could mess with, but the same applies to variable names.

Reply:
Okay, that's slightly better, but still. I don't see why we should have dynamically created list of properties.
It's so much safer to just have them as normal properties; I don't ever want to see any weird bug or behaviour if some species object ends up missing one value in the dictionary.

Reply:
That's the advantage of collections; you can have one statement iterating over every key, and not have a chance of missing one. With properties you need to remember to set each one in its own statement.

Reply:
If you want to have proper design, I'd add an interface IBehaviourProperties or something like that, with an extension method to copy the properties. I really don't think adding the potential class of bugs about dictionary missing keys potentially and all code wanting access needing to go through the dictionary than just using a property, is worth it to make it easier to loop through values.

Reply:
I've already backed out my review. I'm not going to argue with you over poor Fuja's PR.